Backports for 1.1.0-rc2

KristofferC opened the pull request:
Backported PRs:
- [x] #30483 - make inference of `_apply` match the implementation more closely
- [x] #30470 - use type inference world in `return_type`
- [x] #30599 - Use XCode 8.3 for macOS on Travis
- [x] #30598 - Fix use counts for mutable struct SROA
- [x] #30566 - some latency hacks
- [x] #30601 - Fix #29983
- [x] #30666 - Force specialization of the SubArray boundserror method.
- [x] #30497 - Add BB downloading for OpenBLAS
- [x] #30535 - Fix `bb-install` naming conventions, add hashes

Need manual backport:
- [ ] #30560 - Remove duplicate NEWS entry

Non-merged PRs with backport label:
- [ ] #30431 - Build: fix output directory when specified with O=
- [ ] #30403 - Include @​propagate_inbounds docstring in documentation
- [ ] #30398 - Fix enumerate documentation
- [ ] #30367 - move generate_precompile to Base so that it is bundled in downloads
- [ ] #30357 - workaround for F_DUPFD_CLOEXEC on Linux kernel <2.6.24 before it existed

staticfloat committed:
Add BB downloading for OpenBLAS (#30497)

* Auto-detect binarybuilder triplet

* Add OpenBLAS BinaryBuilder installation scaffolding

Also make it easier to add more BB-cached versions of dependencies in the future

* Enable `fixup-libgfortran.sh` to directly ask `$FC` for paths

* Tell Appveyor and Travis to use BinaryBuilder OpenBLAS

Also allow the build system to auto-guess the triplet

(cherry picked from commit 87c18d88aefed9154af18aea9cdabca76763409e)

JeffBezanson committed:
don't use abstract_iteration for types with special cases in _apply (#30483)

(cherry picked from commit 3e6f60758fd9b974972a78bbe899bb754e97a533)

JeffBezanson committed:
use type inference world in `return_type` (#30470)

(cherry picked from commit a5a58828dffbd70b4c9bda3b35dbdd1e324cfe85)

staticfloat committed:
Fix `bb-install` naming conventions, add hashes (#30535)

* Fix `bb-install` naming conventions, add hashes

* Set `DEP_LIBS` to include `openblas` on Appveyor

* When guessing BB libc, default to `glibc` on Linux

* Fix bb-install bash parsing failure

Quote to avoid bash freaking out from spurious `)` from compilers that
have fancy version strings such as `(Red Hat 7.3.1-5)`, which gets
`lastword`'ed down to `7.3.1-5)`.

* Add `contrib/refresh_bb_tarballs.sh` to aid in batch-grabbing BB hashes

(cherry picked from commit 8b189ec494479d9e2ffda0549f066ba2c6389182)

Keno committed:
Fix use counts for mutable struct SROA

PR #28478 moved the computation of the use counts before the finish call.
to fix #28444. However, the early parts of the finish call fixes up phi
node arguments, which fail to get counted if we look at use counts before
that fixup is performed. This causes #30594 where the only non-trivial use
is on the backedge of the phi and would thus incorrectly fail to get accounted
for. Fix that by taking the use count after phi fixup but before dce.

(cherry picked from commit f8f20453c4ca7a6aeb23a189a420dbff130679ce)

Keno committed:
Defensively fix patterns similar to #29983

I don't have concrete tests for these, but it looks like
they all need the `is_old` predicate for what they're doing,
so switch those over also while we're at it.

(cherry picked from commit 34f7a4a50458594e6d72793b4f371227e8e842e3)

ararslan committed:
Use XCode 8.3 for macOS on Travis (#30599)

This corresponds to macOS 10.12 Sierra. XCode 8 covers El Capitan and
Sierra, so if Travis is giving us XCode 8.x for x < 3, we're on El Cap.
Homebrew supports only three versions of macOS at a time, which means
that El Cap (10.11) is no longer support. This is likely why our Mac
builds are trying to build GCC from source; a bottle might not be
available.

(cherry picked from commit 862fe08de4283346b51e8dcddb3df52a5315ac6a)

ararslan commented:
I've marked #30599 for backporting, which should fix the macOS CI. @staticfloat, it looks like we'll also need to backport whatever it is you did that fixed AppVeyor.

staticfloat commented:
So probably https://github.com/JuliaLang/julia/commit/8b189ec494479d9e2ffda0549f066ba2c6389182 then

Keno committed:
Fix SROA confusing new and old nodes

SROA was accidentally treating a pending node as old and thus
getting the wrong type when querying the predecessor. As a result
it thought one of the paths was unreachable causing undefined data
to be introduced on that path (generally the `1.0` that happened to
already be in register).

Fix #29983

(cherry picked from commit da0179c4d60c0fa3e5d64c08972773c82d57e4e6)

JeffBezanson committed:
some latency hacks (#30566)

These changes cut some method dependency edges that tend to lead to
invalidations when loading packages.

(cherry picked from commit 13fc4c39a37742b8f804c5268426c49c7b31b7f5)

maleadt committed:
Force specialization of the SubArray boundserror method.

Improves https://github.com/JuliaLang/julia/pull/29867 by avoiding an invoke.

(cherry picked from commit f2d2d6f465da338062fee79bddf18a9cf56652fd)

KristofferC commented:
@nanosoldier `runbenchmarks("linalg", vs = ":release-1.1")`

KristofferC commented:
@ararslan could you kick nanosoldier?

staticfloat commented:
The appveyor failure is bizarre; `perl` is complaining that it can't make changes without a backup here: https://github.com/JuliaLang/julia/blob/346199d100ea106fa0f80a33ee4a3333a605a273/deps/blas.mk#L96

That line hasn't changed in 3 years.  Could `perl` have changed out from underneath us?

staticfloat commented:
Also why are we even using `perl` here; we could just turn that into a normal `.patch`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/30607#issuecomment-452829815) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d957a6949aac37a6de1e345f2967c7a329ccc6ce/dae984d_vs_ecf9911/report.md). cc @ararslan

vtjnash commented:
> Also why are we even using perl here; we could just turn that into a normal .patch.

We could try, but this bug has been fixed since v0.1alpha2.5 over 7 years ago, so it would just be an empty file. (https://github.com/xianyi/OpenBLAS/commit/0d76196a099991b199b286c932b1a9919bcd0a48)

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs=":release-1.1")`

KristofferC committed:
remove unneded patching of Makefile

(cherry picked from commit 2a2ba37bf29ff73fc32f6df1c1ab79d569d1d7e2)

KristofferC commented:
I tried to just remove it.

staticfloat commented on deps/blas.mk:
```diff
@@ -92,10 +92,6 @@ endif
 OPENBLAS_BUILD_OPTS += MAKE_NB_JOBS=0
 
-$(BUILDDIR)/$(OPENBLAS_SRC_DIR)/build-configured: $(BUILDDIR)/$(OPENBLAS_SRC_DIR)/source-extracted
-	perl -i -ple 's/^\s*(EXTRALIB\s*\+=\s*-lSystemStubs)\s*$$/# $$1/g' $(dir $<)/Makefile.system
```
I think you still need to create the `build-configured` file though; so just remove the `perl` invocation

KristofferC replied:
Alright,  thanks.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/30607#issuecomment-452899345) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/43ac38d25687f578eb62b74cc9ec22d540cc9a83/dae984d_vs_ecf9911/report.md). cc @ararslan

KristofferC commented:
AV seems to build OpenBLAS from scratch now and times out during test on 32 bit but fails on 64 bit with:

```
cp: cannot stat 'scratch/openblas-fd8d1868a126bb9f12bbc43b36ee30d1ba943fbb/libopenblas64_.dll': No such file or directory
make[1]: *** [/cygdrive/c/projects/julia/deps/blas.mk:112: /cygdrive/c/projects/julia/usr-staging/openblas-fd8d1868a126bb9f12bbc43b36ee30d1ba943fbb.tgz] Error 1
make[1]: Leaving directory '/cygdrive/c/projects/julia/deps'
make: *** [Makefile:57: julia-deps] Error 2
```

Did I do something bad in the last commit? Help plz.

ararslan commented:
PkgEval shows no regressions since RC1. Note that it was run prior to the most recent Makefile changes.

staticfloat commented:
I'm not sure how, but it looks like you only got part of 87c18d88aefed9154af18aea9cdabca76763409e.  You need the rest of the changes as well.

KristofferC commented:
Well, that wasn't marked with backport label and it wasn't the commit in  https://github.com/JuliaLang/julia/pull/30607#issuecomment-451718062 so it wasn't backported. I'll add that as well then. 

But why do we need so many build backports to this branch? Did something change on AV's end?

KristofferC commented:
Backported those two but something is still wrong :(

```
make[1]: Leaving directory '/cygdrive/c/projects/julia'
make[1]: Entering directory '/cygdrive/c/projects/julia'
fatal: cannot change to '/cygdrive/c/projects/julia': No such file or directory
```

staticfloat commented:
I have no idea what Appveyor is smoking there.  That doesn't make any sense to me, especially as we can see other processes happily going and entering that directory.

KristofferC commented:
Ok, this happens on master as well. Doesn't keep the tests from running apparently.
